bazel_tools/scala.bzl

```diff
@@ -201,6 +201,121 @@ def _create_scala_source_jar(**kwargs):
+            "-doc-title",
+            ctx.attr.doctitle,
+            "-no-link-warnings",
+        ] + common_scalacopts + srcFiles,
+        mnemonic = "ScaladocGen",
+    )
+
+    # since we only have the output directory of the scaladoc generation we need to find
+    # all the files below sources_out and add them to the zipper args file
+    zipper_args_file = ctx.actions.declare_file(ctx.label.name + ".zipper_args")
+    ctx.actions.run_shell(
+        mnemonic = "ScaladocFindOutputFiles",
+        outputs = [zipper_args_file],
+        inputs = [outdir],
+        command = "find -L {src_path} -type f | sed -E 's#^{src_path}/(.*)$#\\1={src_path}/\\1#' | sort > {args_file}".format(
+            src_path = outdir.path,
+            args_file = zipper_args_file.path,
+        ),
+        progress_message = "find_scaladoc_output_files %s" % zipper_args_file.path,
+        use_default_shell_env = True,
+    )
+
+    ctx.actions.run(
+        executable = ctx.executable._zipper,
+        inputs = ctx.files.srcs + classpath + [outdir, zipper_args_file],
+        outputs = [ctx.outputs.out],
+        arguments = ["c", ctx.outputs.out.path, "@" + zipper_args_file.path],
+        mnemonic = "ScaladocJar",
+    )
+
+scaladoc_jar = rule(
+    implementation = _scaladoc_jar_impl,
+    attrs = {
+        "deps": attr.label_list(),
+        "doctitle": attr.string(default = ""),
+        "plugins": attr.label_list(default = []),
+        "srcs": attr.label_list(allow_files = True),
+        "_zipper": attr.label(
+            default = Label("@bazel_tools//tools/zip:zipper"),
+            cfg = "host",
+            executable = True,
+            allow_files = True,
+        ),
+        "_scaladoc": attr.label(
+            default = Label("@scala_nix//:bin/scaladoc"),
+            cfg = "host",
+            executable = True,
+            allow_files = True,
+        ),
+    },
+    outputs = {
+        "out": "%{name}.jar",
+    },
+)
+"""
+Generates a Scaladoc jar path/to/target/<name>.jar.
+
+Arguments:
+  srcs: source files to process
+  deps: targets that contain references to other types referenced in Scaladoc.
+  doctitle: title for Scalaadoc's index.html. Typically the name of the library
+"""
+
+def _create_scaladoc_jar(**kwargs):
+    # Try to not create empty scaladoc jars and limit execution to Linux and MacOS
+    # Detect an actual scala source file rather than a srcjar or other label
+    if len(kwargs["srcs"]) > 0 and is_windows == False:
+        for src in kwargs["srcs"]:
+            if src.endswith(".scala"):
+                plugins = []
+                if "plugins" in kwargs:
+                    plugins = kwargs["plugins"]
+
+                scaladoc_jar(
+                    name = kwargs["name"] + "_scaladoc",
+                    deps = kwargs["deps"],
+                    plugins = plugins,
+                    srcs = kwargs["srcs"],
+                )
+                break
```

Comment:
I think it would be better to have the loop set a boolean, and then conditionally call `scaladoc_jar` outside the `for src ...` loop. Currently I find it a bit hard to read. It took me a while to realize that you don't call `scaladoc_jar` for each source file. E.g.

```
create_scaladoc = False
for src in kwargs["srcs"]:
    if src.endswith(".scala"):
        create_scaladoc = True
        break
if create_scaladoc:
    ...
```

Reply:
Also a good idea. Implemented.